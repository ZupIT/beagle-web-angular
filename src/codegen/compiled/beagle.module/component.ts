/*
  * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  *  http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
*/

import { Type } from '@angular/core'
import {
  kebabToCamelCase,
  removeExtraIndentation,
  replaceToUnderline,
} from '../../utils/formatting'
import { getComponentAnnotations } from '../../utils/metadata'
import { remoteViewSelector } from '../../../constants'


function createViewChildString(name: string, templateName: string, angularVersion: number): string {
  const componentName = replaceToUnderline(name)

  if (angularVersion >= 8) {
    return `'${componentName}': new ViewChild('${templateName}', { static: true })`
  } else {
    return `'${componentName}': new ViewChild('${templateName}')`
  }
}

function createQueries(components: Record<string, Type<any>>, angularVersion: number) {
  const componentNames = Object.keys(components)

  return componentNames.map((name) => {
    const selector = getComponentAnnotations(components[name]).selector
    const templateName = kebabToCamelCase(selector)
    return createViewChildString(name, templateName, angularVersion)
  })
}


export function createComponentString(
    components: Record<string, Type<any>>,
    angularVersion: number,
  ) {
  const queries = createQueries(components, angularVersion)

  // todo: legacy code. Remove the input "loadParams" with v2.0.
  const componentString = `
    @Component({
      selector: '${remoteViewSelector}',
      template,
      inputs: ['loadParams', 'route', 'networkOptions', 'controllerId'],
      queries: {
        ${queries.join(',\n        ')},
      },
    })
    export class BeagleRemoteView extends AbstractBeagleRemoteView {
      constructor(
        beagleProvider: BeagleProvider,
        ngZone: NgZone,
        changeDetector: ChangeDetectorRef,
      ) {
        // @ts-ignore
        super(beagleProvider, ngZone, changeDetector)
      }
    }
  `

  return removeExtraIndentation(componentString, 4)
}
