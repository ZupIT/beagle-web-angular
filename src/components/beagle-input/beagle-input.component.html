<!-- 
  * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  *  http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License. 
-->

<div class="wrapper">
  <input
    [type]="type"
    [placeholder]="placeholder"
    [attr.disabled]="disabled ? '' : null"
    [readOnly]="readOnly"
    [hidden]="hidden"
    [(ngModel)]="value"
    [value]="value"
    (ngModelChange)="handleChange($event)"
    (blur)="handleBlur($event)"
    (focus)="handleFocus($event)" 
    [ngClass]="{'hasError': error && showError, '_beagleFormError': error}"
    />

  <beagle-text *ngIf="error && showError" text={{error}} textColor="red"></beagle-text>
</div>