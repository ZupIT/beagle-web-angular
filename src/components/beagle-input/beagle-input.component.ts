/*
  * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  *  http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
*/

import { Component, Input, OnInit, ViewEncapsulation } from '@angular/core'
import { BaseComponent } from '../../runtime/BaseComponent'
import { BeagleInputInterface, InputType } from '../schemas/input'
import { InputHandler } from '../schemas/input-handler'

@Component({
  selector: 'beagle-input',
  templateUrl: './beagle-input.component.html',
  styleUrls: ['./beagle-input.component.less'],
  encapsulation: ViewEncapsulation.None,
})
export class BeagleInputComponent extends BaseComponent 
  implements BeagleInputInterface, OnInit {

  @Input() value?: string
  @Input() placeholder?: string
  @Input() disabled?: boolean
  @Input() readOnly?: boolean
  @Input() type?: InputType
  @Input() onChange?: InputHandler
  @Input() onBlur?: InputHandler
  @Input() onFocus?: InputHandler
  @Input() error?: string
  @Input() showError?: boolean

  ngOnInit() {
    this.value = this.value || ''
    this.placeholder = this.placeholder || ''
    this.disabled = this.disabled || false
    this.readOnly = this.readOnly || false
    this.type = this.type || 'TEXT'
    this.showError = this.showError || false
  }

  public handleChange(value) {
    this.onChange && this.onChange({ value })
  }

  public handleBlur(event) {
    this.onBlur && this.onBlur({ value: event.target.value })
  }

  public handleFocus(event) {
    this.onFocus && this.onFocus({ value: event.target.value })
  }
}
