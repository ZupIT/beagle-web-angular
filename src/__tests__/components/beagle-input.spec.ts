/*
  * Copyright 2020 ZUP IT SERVICOS EM TECNOLOGIA E INOVACAO SA
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
  * you may not use this file except in compliance with the License.
  * You may obtain a copy of the License at
  *
  *  http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
*/

import { TestBed, async, ComponentFixture, tick, fakeAsync } from '@angular/core/testing'
import { FormsModule } from '@angular/forms'
import { By } from '@angular/platform-browser'
import { BeagleInputComponent } from '../../components/beagle-input/beagle-input.component'
import { buildBeagleTestModuleMetadata } from './test-module-builder'
import { BeagleTextComponent } from '../../components/beagle-text/beagle-text.component'

let component: BeagleInputComponent
let fixture: ComponentFixture<BeagleInputComponent>


function setAndCallHandler(selector: string, value: string, event: string) {
  fixture.detectChanges()
  tick()

  const input = fixture.debugElement.query(By.css('input')).nativeElement
  input.value = value
  input.dispatchEvent(new Event(event))
  tick()
}

describe('BeagleInputComponent', () => {

  beforeEach(async(() => {
    TestBed.configureTestingModule(
      buildBeagleTestModuleMetadata([BeagleInputComponent, BeagleTextComponent], [FormsModule]))
      .compileComponents()

    fixture = TestBed.createComponent(BeagleInputComponent)
    component = fixture.componentInstance
    component.onFocus = jest.fn()
    component.onChange = jest.fn()
    component.onBlur = jest.fn()
    component.error = 'Jest test',
    component.showError = false
  }))

  it('should match snapshot', () => {
    expect(component).toMatchSnapshot()
  })

  it('should call handlers and set Inputs', fakeAsync(() => {
    spyOn(component, 'handleChange').and.callThrough()
    spyOn(component, 'handleFocus').and.callThrough()
    spyOn(component, 'handleBlur').and.callThrough()

    setAndCallHandler('input', 'Testing', 'input')
    setAndCallHandler('input', 'Testing', 'focus')
    setAndCallHandler('input', 'Testing', 'blur')

    expect(component.onChange).toHaveBeenCalled()
    expect(component.onFocus).toHaveBeenCalled()
    expect(component.onBlur).toHaveBeenCalled()
  }))

  it('should call handlers and set Inputs', fakeAsync(() => {
    component.value = 'New input text'
    const nativeInput = fixture.debugElement.nativeElement.querySelector('input')
    fixture.detectChanges()
    expect(nativeInput.value).toBe('New input text')
  }))
})
