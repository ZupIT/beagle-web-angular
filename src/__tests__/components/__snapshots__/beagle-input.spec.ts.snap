// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BeagleInputComponent should match snapshot 1`] = `
BeagleInputComponent {
  "onBlur": [MockFunction],
  "onChange": [MockFunction],
  "onFocus": [MockFunction],
}
`;